Let players use the console during FMVs

rr- opened the pull request:
#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TRX/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [x] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description

This builds on top of #3453 to let the players interact with the dev console even when an FMV is playing. 
Also, resolves #3147.

#### Technical considerations and testing

In develop we do not have the font injected until we actually load the title inventory ring. To circumvent this, I changed the game flow sequencer to always load the level at the start of a sequence (unless we're playing "story so far", then it'll skip loading the level file for non-cutscene levels for performance reasons), and merged TR1 and TR2 loading routines. As such it'd be good to test that this reordering didn't break anything (especially when it comes to the inventory management):

- Save/load
- Select level (TR1)
- Play any level (TR2)
- Restart level × new game/savegame
- Story so far
- Secret rewards in TR2
- Inventory modifiers
    - Starting without pistols in Natla's Mines
    - Saving with pistols in Natla's Mines and reloading the game does not remove pistols
    - Advancing to The Great Pyramid without pistols gives Lara pistols
    - Saving and reloading in The Great Pyramid doesn't mess with Lara's items
- Save crystals continue to ask for a save at a start of a level
- Save/load dialog row preselection works as expected
- Bonus levels trigger as expected and do not trigger as expected
- Cutscenes shouldn't crash, and Lara's braid should operate as normal

rr- committed:
tr1/fmv: add ability to use console

rr- committed:
game-flow: load levels at sequence start

This lets the level machinery inject O_ALPHABET before the first FMV
starts playing, letting the player interact with the console and see the
characters even if they open the console prior to showing the title
screen.

rr- committed:
game-flow: fix inventory management with /play

Resolves #3147.

rr- commented:
Added a fix to 3147 from Lahm.
Both games were affected by it, but in TR1 it only would manifest with a custom game flow.

rr- committed:
tr2/fmv: add ability to use console

rr- commented:
Update: console is now usable in TR2 FMVs as well.

lahm86 commented on src/libtrx/game/game_flow/sequencer.c:
```diff
@@ -78,6 +82,101 @@ GF_COMMAND GF_InterpretSequence(
+        } else {
+            // console /play level feature
+            Savegame_InitCurrentInfo();
+            Inv_RemoveAllItems();
```
We just need an `#include` for this, current compiler warning. My bad, should have sent the full diff.

rr- replied:
:+1: added

lahm86 commented:
- The title music starts as soon as the game starts, then is cut off by the initial FMVs, and starts again on the title screen.
- If you load a level that has an FMV, there is no ambient music until you save/load.
- A side effect is you can issue commands for the level during the FMV, like `/kill all` or `/tp`. Not so much an issue, but perhaps we'd want to thwart those.

Just an idea to throw out there - what if we shipped a dummy level file (basically empty) and then load that during FMVs, so allowing us to inject? Edit: or perhaps just using `title.phd`?

aredfan commented:
Issues:

- [x] TR2: The special animation at the start of Offshore Rig/HSH get skipped.
- [x] TR1: The game crashes after the final stats screen. UB is unaffected.
- [x] TR2: I think the shotgun issue still persists:
> 1. start up the game
> 2. select Lara's Home
> 3. press F5 to start a new game
> 4.  draw the shotgun and see the ammo counter is empty

rr- commented:
> Just an idea to throw out there - what if we shipped a dummy level file (basically empty) and then load that during FMVs, so allowing us to inject? Edit: or perhaps just using `title.phd`?

I tried that and it was a can of worms much bigger than the current issues.

rr- committed:
level: move Music_Play outside of level init

rr- commented:
- [x] The title music starts as soon as the game starts, then is cut off by the initial FMVs, and starts again on the title screen.
- [x] If you load a level that has an FMV, there is no ambient music until you save/load.
- [x] A side effect is you can issue commands for the level during the FMV, like /kill all or /tp. Not so much an issue, but perhaps we'd want to thwart those.
    - [x] this crashes the game if issued with `/play 1` and then `/kill all` while still in the cutscene, as `g_MatrixPtr` is null
    - [x] check all commands
- [x] TR2: The special animation at the start of Offshore Rig/HSH get skipped.
- [x] TR1: The game crashes after the final stats screen. UB is unaffected.
- [x] Inventory music plays for split second when issuing `/play 1` while in the opening fmv
- [x] TR2: I think the shotgun issue still persists:
    - start up the game
    - select Lara's Home
    - press F5 to start a new game
    - draw the shotgun and see the ammo counter is empty
- [x] TR2 loading a savegame crashes the game complaining about holster meshes.
- [x] TR1 loading a savegame crashes the game…
- [x] Lara pulls out pistols at the start of every level (………)
- [x] TR1: the game crashes when starting demos

rr- committed:
matrix: always initialise pointer to stack base

rr- committed:
phase: more eager overriding of commands

No longer waits till a control() callback of a phase before issuing an
override command.

This helps with a scenario where the player issues a command like
`/play` during the opening FMV before the inventory ring was shown – now
the phase won't even try to start and as such will omit the music
trigger.

rr- committed:
console/cmd: disable some commands during fmvs

rr- commented:
@aredfan If you don't mind I'd appreciate one more recheck – I've tidied up the commits (not much attention needed here) and – I think – I got the shotgun issue fixed in a pretty solid way.

aredfan commented:
> @aredfan If you don't mind I'd appreciate one more recheck – I've tidied up the commits (not much attention needed here) and – I think – I got the shotgun issue fixed in a pretty solid way.

No problem, the shotgun fix is working great! 👍 

I found a few more issues while retesting. I also made sure to check these things doesn't happen in develop.

- [x]  TR1: Lara has holsters when starting the gym.
- [x] TR1: Issuing the teleport command during cutscenes crashes the game:

> 1. `/cut 1`
> 2. `/tp 10`

- [x] TR2: Lara doesn't draw pistols when the draw input is pressed in some levels.  The pistols will need to be manually selected in the inventory.

> 1. `/play` 6-17 or `/play 5` in GM
> 2. press the draw input

rr- commented:
Thanks – fixed all issues :relieved:

aredfan commented:
> Thanks – fixed all issues 😌

Thank you again.  I can confirm the cutscene crash and draw pistols issues are fixed.  I'm afraid Lara does still have her holsters in the TR1 gym.

<details><summary>Screenshot</summary>

![20250707_110715_Lara's_Home](https://github.com/user-attachments/assets/dafe5514-a5b5-4a1f-86f9-98c5fe1aa6ac)

</details>

rr- commented:
@aredfan I forgot to push the changes before leaving the house :sweat:  should be good now
![20250707_123558_Lara's_Home](https://github.com/user-attachments/assets/45761824-2098-4cfd-952c-9ceafb7f18c5)

aredfan commented:
Thank you @rr-, the holsters are still appearing on my end, so it made me wonder whether there's another reason the bug is manifesting.  Then it hit me, I'm manually selecting Lara's Home which itself leads to the holsters appearing, whereas `/play 0` doesn't make the holsters appear. 💡

(Confirmed in your screenshot `Loading Lara's Home`)

rr- commented:
> Thank you @rr-, the holsters are still appearing on my end, so it made me wonder whether there's another reason the bug is manifesting. Then it hit me, I'm manually selecting Lara's Home which itself leads to the holsters appearing, whereas `/play 0` doesn't make the holsters appear. 💡
> 
> (Confirmed in your screenshot `Loading Lara's Home`)

Oh I see! Fixed this as well.

The fix can affect starting the game in normal circumstances, ending a level, and playing bonus levels.

aredfan reviewed (approved):
> The fix can affect starting the game in normal circumstances, ending a level, and playing bonus levels.

I retested with manually selecting the gym and starting new game, and did level skipping to the credits in both games. I also tested GM with the bonus locked and unlocked.

I think it all looks good. 😌

lahm86 commented on src/tr1/game/fmv.c:
```diff
@@ -85,6 +87,9 @@ static void M_UploadSurface(void *const surface, void *const user_data)
     UI_BeginScene();
+    Console_Draw();
+    Console_Control();
+    Console_Control();
```
Sorry, just noticed this. Is the double control call needed? Same seems to be in TR2 below.

rr- replied:
It's a necessary hack to have the caret blink at roughly the right frame rate as technically FMV player is not a phaser, and uses its own event loop.
